ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3BatchDeleteHandler.java

```diff
@@ -0,0 +1,183 @@
+          CountDownLatch latch = new CountDownLatch(1);
+
+          // Handle the delete operation using the deleteBlobHandler
+          deleteBlobHandler.handle(singleDeleteRequest, noOpResponseChannel, new Callback<Void>() {
+            @Override
+            public void onCompletion(Void result, Exception exception) {
+              // Call our custom onDeleteCompletion to track success/failure
+              Map<String, Object> args = restRequest.getArgs();
+              restRequest.getArgs().remove("ambry-internal-key-target-account");
+              restRequest.getArgs().remove("ambry-internal-key-keep-alive-on-error-hint");
+              restRequest.getArgs().remove("ambry-internal-key-target-container");
+              boolean success = exception == null;
+              onDeleteCompletion(success, object.getKey());
+
+              // Decrement latch to signal that this delete operation has finished
+              latch.countDown();
+            }
+          });
+
+          // Wait until all deletes are completed
+          try {
+            latch.await();
+          } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+          }
+        }
+
+        // Check if any delete request failed
+        if (failedRequest) {
+          restResponseChannel.setStatus(ResponseStatus.BadRequest);
+          callback.onCompletion(null, new RestServiceException("Failed to execute S3BatchDelete", RestServiceErrorCode.BadRequest));
+        } else {
+          // Successful response handling
+          restResponseChannel.setStatus(ResponseStatus.Ok);
+          try {
+            XmlMapper xmlMapper = new XmlMapper();
+            S3MessagePayload.S3BatchDeleteResponse resp = new S3MessagePayload.S3BatchDeleteResponse();
+            resp.setDeleted(deleted);
+            resp.setErrors(errors);
+            ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+            xmlMapper.writeValue(outputStream, resp);
+            ReadableStreamChannel readableStreamChannel =
+                new ByteBufferReadableStreamChannel(ByteBuffer.wrap(outputStream.toByteArray()));
+            restResponseChannel.setHeader(RestUtils.Headers.CONTENT_LENGTH, readableStreamChannel.getSize());
+            restResponseChannel.setHeader(RestUtils.Headers.CONTENT_TYPE, XML_CONTENT_TYPE);
+            callback.onCompletion(readableStreamChannel, null);
+          } catch (IOException e) {
+            callback.onCompletion(null, new RestServiceException("Failed to serialize response", RestServiceErrorCode.InternalServerError));
+          }
+        }
+      }
+    }, restRequest.getUri(), LOGGER, callback);
+  }
+
+  /**
+   * This method is used to update the lists of deleted and errored objects.
+   * @param success whether the delete was successful or not
+   * @param key the object key
+   */
+  public void onDeleteCompletion(boolean success, String key) {
```

Comment:
This fn can be folded into the call site. It's small and not called anywhere else.